service/history/dlq.go

```diff
@@ -0,0 +1,72 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package history
+
+import (
+	"go.uber.org/fx"
+
+	"go.temporal.io/server/common/cluster"
+	"go.temporal.io/server/common/dynamicconfig"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/service/history/queues"
+)
+
+type (
+	executableDLQWrapper struct {
+		*executableDLQWrapperParams
+	}
+	executableDLQWrapperParams struct {
+		fx.In
+
+		HistoryTaskQueueManager persistence.HistoryTaskQueueManager
+		ClusterMetadata         cluster.Metadata
+		DC                      *dynamicconfig.Collection
+	}
+	executableToggle struct {
+		queues.Executable
+		executableDLQ *queues.ExecutableDLQ
+		useDLQ        dynamicconfig.BoolPropertyFn
+	}
```

Comment:
do you plan to use this struct for other functionalities in the future? If so, it should probably live in a dedicated file, instead of dlq.go. If no, we can move the dc into ExecuableDLQ?

Reply:
I want this to be specific to the main entry point because the other unit tests don't need this. Basically, in general, I think feature flags should be decoupled from the feature itself and just wrap them in the main entry point so that we don't end up with a bunch of places in tests where we have to do something like `dcOverrides: map[string]interface{}{ dynamicconfig.HistoryTaskDLQEnabled: true }`. I also think this should only be a temporary thing because we probably want the DLQ on always once it's tested.

Reply:
synced offline, plz rename to executableDLQToggle or something similar.